File: anki/explain.py
#!/usr/bin/env python3
"""Explain one English word for a Korean learner, and cache the answer.

A dictionary says what a word means. It does not say whether you would use it in
a job application or to a friend, which words it habitually travels with, or
which near-identical word you are about to confuse it with. Those are the things
that decide whether a word is actually usable, and they are what this asks for.

Answers are written to a cache in the repository, keyed by the word. Looking the
same word up again is then free and instant, which over a year of study is most
lookups.

    ANTHROPIC_API_KEY, ANTHROPIC_MODEL (optional)
"""

from __future__ import annotations

import argparse
import json
import os
import re
import sys
from pathlib import Path

from pydantic import BaseModel

from proofread import MODEL, api_key, client, key_shape  # noqa: E402

SYSTEM = """You explain English words to a Korean adult who is building \
vocabulary flashcards. They are past beginner: they know the common words and \
are working on the ones that make writing sound native.

Explain the word as given. If it is a phrase or a phrasal verb, explain the \
phrase, not its parts.

- meaning: the plain English definition, one line. Say what it means before \
saying anything clever about it.
- pronunciation: IPA, in slashes, as a British or American speaker would say it.
- korean: the most accurate Korean gloss. If no single word fits, give the \
short phrase that does. Do not list five near-synonyms.
- tone: the register, as one of formal, neutral, casual, slang, literary, \
technical. Say which, and nothing else.
- nuance: two sentences at most. What this word carries that a plainer synonym \
does not, and when a native speaker would reach for it. Concrete, not abstract.
- collocations: the words it habitually travels with, written as real \
fragments — "a resilient economy", "bounce back from". Three to five. These \
matter more than the definition for sounding natural.
- confusables: words this learner is likely to reach for by mistake, each with \
the one distinction that separates them. Two at most. Skip it if nothing \
genuinely competes.
- examples: two sentences a real person would say, at this learner's level, \
showing different senses or registers if the word has them.
- memory_hook: one line. An etymology, a cognate, or an image that makes it \
stick. Skip it rather than force something weak.

Be specific and brief. This goes on a flashcard that will be read hundreds of \
times, so every word has to earn its place."""


class Confusable(BaseModel):
    word: str
    difference: str


class Explanation(BaseModel):
    word: str
    meaning: str
    pronunciation: str
    korean: str
    tone: str
    nuance: str
    collocations: list[str]
    confusables: list[Confusable]
    examples: list[str]
    memory_hook: str


def slug(term: str) -> str:
    """Cache key. Stable, lowercase, and safe as a filename."""
    return re.sub(r"[^a-z0-9]+", "-", term.strip().lower()).strip("-")


def explain(term: str, client, *, model: str = "") -> Explanation:
    response = client.messages.parse(
        model=model or MODEL,
        max_tokens=4000,
        system=SYSTEM,
        messages=[{"role": "user", "content": term.strip()}],
        output_format=Explanation,
    )
    if getattr(response, "stop_reason", None) == "refusal":
        raise RuntimeError("the model declined to explain this word")
    return response.parsed_output


def main() -> int:
    parser = argparse.ArgumentParser(description="Explain a word and cache the answer.")
    parser.add_argument("--word", required=True)
    parser.add_argument("--cache-dir", default="docs/lookups")
    parser.add_argument("--model", default="")
    parser.add_argument("--force", action="store_true", help="Re-ask even if it is cached.")
    args = parser.parse_args()

    term = args.word.strip()
    if not term:
        print("::error::No word was given.", flush=True)
        return 1
    if len(term) > 80:
        print("::error::That is too long to be a word or a phrase.", flush=True)
        return 1

    key = slug(term)
    if not key:
        print(f"::error::{term!r} has nothing to look up in it.", flush=True)
        return 1

    target = Path(args.cache_dir) / f"{key}.json"
    if target.exists() and not args.force:
        print(f"Already cached at {target}; nothing to do.", flush=True)
        return 0

    import anthropic

    # Without this the SDK raises a TypeError about header resolution, which
    # says nothing about what to actually do. The fix is one setting, so name it.
    if not api_key():
        print(
            "::error::ANTHROPIC_API_KEY is not set for this repository. "
            "Add it under Settings -> Secrets and variables -> Actions -> "
            "New repository secret, named exactly ANTHROPIC_API_KEY. A secret on "
            "your account or on an environment is not the same thing and this "
            "workflow cannot see it.",
            flush=True,
        )
        return 1

    print(f"Asking about {term!r}...", flush=True)
    try:
        answer = explain(term, client(), model=args.model)
    except anthropic.AuthenticationError:
        print(
            f"::error::The API key was rejected: {key_shape()}. Replace the "
            "ANTHROPIC_API_KEY secret with a key copied whole from "
            "console.anthropic.com. The console shows a key in full only at the "
            "moment you create it; afterwards it is masked, and the masked form "
            "is not a usable key.",
            flush=True,
        )
        return 1
    except anthropic.BadRequestError as exc:
        note = str(exc).lower()
        if "workspace" in note:
            print(
                "::error::The key works, but it belongs to the organization rather "
                "than to any one workspace, and the API will not pick a workspace "
                "for it. Two ways out, either is fine: make a new key from inside a "
                "workspace at console.anthropic.com (it carries the workspace with "
                "it, and nothing else needs changing), or add the workspace's ID as "
                "a repository variable named ANTHROPIC_WORKSPACE_ID and this will "
                "send it.",
                flush=True,
            )
            return 1
        if "credit balance" in note:
            print(
                "::error::The key works, but the account has no credit. Buy credit "
                "at console.anthropic.com -> Billing. This is separate from a "
                "Claude subscription; a Pro or Max plan does not pay for API calls.",
                flush=True,
            )
            return 1
        raise

    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(
        json.dumps(answer.model_dump(), ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
    )
    print(f"Wrote {target}", flush=True)

    summary = os.environ.get("GITHUB_STEP_SUMMARY")
    if summary:
        lines = [
            f"## {answer.word}", "",
            f"**{answer.korean}** · _{answer.tone}_ · `{answer.pronunciation}`", "",
            answer.meaning, "",
            answer.nuance, "",
            "**Goes with:** " + ", ".join(f"`{c}`" for c in answer.collocations),
        ]
        if answer.confusables:
            lines += ["", "**Not to be confused with:**"] + [
                f"- **{c.word}** — {c.difference}" for c in answer.confusables
            ]
        lines += ["", "**Examples:**"] + [f"- {e}" for e in answer.examples]
        if answer.memory_hook:
            lines += ["", f"**Remember it:** {answer.memory_hook}"]
        with open(summary, "a", encoding="utf-8") as handle:
            handle.write("\n".join(lines) + "\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: docs/app/explain.js
// Why a word means what it means, at the moment it did not stick.
//
// A definition tells you what a word means. It does not tell you when a native
// speaker reaches for it, what it habitually travels with, or which
// near-identical word you were probably confusing it with — and those are what
// a card that keeps failing is failing on. So this is shown on the answer side,
// and shown by itself after an Again, because that is the only moment when
// someone is reliably willing to read it.
//
// Cheapest source first, and most of the time the cheapest is free:
//
//   1. this browser          — instant, already paid for
//   2. the deployed site     — a static file in the repository, no key, no cost
//   3. the API               — only for a word nobody has asked about yet
//
// The same answers the card form writes are the ones read here, so a word
// explained once anywhere is explained everywhere, forever.

const CACHE = "lexis:why:";
const SITE = "../lookups/";

export function slug(term) {
  return String(term || "").trim().toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "");
}

function remembered(key) {
  try {
    const raw = localStorage.getItem(CACHE + key);
    return raw ? JSON.parse(raw) : null;
  } catch { return null; }
}

function remember(key, info) {
  try { localStorage.setItem(CACHE + key, JSON.stringify(info)); } catch { /* full or blocked */ }
}

async function fromSite(key) {
  try {
    const res = await fetch(`${SITE}${key}.json`, { cache: "no-cache" });
    if (res.ok) return await res.json();
  } catch { /* offline, or not deployed yet */ }
  return null;
}

// The prompt and schema come from the file anki/explain.py writes, so the app
// and the workflow cannot drift into asking for different things.
let contract = null;
async function theContract() {
  if (contract) return contract;
  const res = await fetch("../explain-contract.json", { cache: "no-cache" });
  if (!res.ok) throw new Error("The explanation contract could not be loaded.");
  contract = await res.json();
  return contract;
}

async function fromApi(term, key, apiKey) {
  const { system, schema, max_tokens } = await theContract();
  const res = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-api-key": apiKey,
      "anthropic-version": "2023-06-01",
      "anthropic-dangerous-direct-browser-access": "true",
    },
    body: JSON.stringify({
      model: "claude-haiku-4-5",
      max_tokens,
      system,
      messages: [{ role: "user", content: term }],
      output_config: { format: { type: "json_schema", schema } },
    }),
  });

  const payload = await res.json().catch(() => ({}));
  if (!res.ok) {
    const said = payload?.error?.message || `${res.status}`;
    throw new Error(
      res.status === 401 ? "Claude rejected the key in Settings."
      : res.status === 429 ? "Claude is rate limiting; it will work again shortly."
      : /workspace/i.test(said) ? "That key belongs to the organisation rather than a workspace."
      : /credit balance/i.test(said) ? "The Anthropic account is out of credit."
      : said);
  }
  if (payload.stop_reason === "refusal") throw new Error("Claude declined to explain that one.");
  const text = (payload.content || []).filter((b) => b.type === "text").map((b) => b.text).join("");
  if (!text) throw new Error("Claude sent nothing back.");
  const info = JSON.parse(text);
  if (info.recognised === false) throw new Error(info.meaning || "That does not look like an English word.");
  return info;
}

export async function explain(term, { apiKey = "" } = {}) {
  const key = slug(term);
  if (!key) throw new Error("Nothing to look up.");

  const local = remembered(key);
  if (local) return { info: local, from: "this browser" };

  const site = await fromSite(key);
  if (site) { remember(key, site); return { info: site, from: "the deck" }; }

  if (!apiKey) {
    throw new Error("Not explained yet. Add an Anthropic key under Settings, " +
      "or run the explain workflow for this word.");
  }
  const info = await fromApi(term, key, apiKey);
  remember(key, info);
  return { info, from: "Claude" };
}

function esc(text) {
  return String(text ?? "").replace(/[&<>"]/g, (c) =>
    ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;" }[c]));
}

// The card already shows the word, the meaning and an example. Repeating them
// here would push the part worth reading off the screen, so only what the card
// cannot say is rendered.
export function render(info, from) {
  let html = "";
  if (info.nuance) html += `<h4>when you would use it</h4><p>${esc(info.nuance)}</p>`;

  if ((info.collocations || []).length) {
    html += `<h4>goes with</h4><div>` +
      info.collocations.map((c) => `<code>${esc(c)}</code>`).join("") + `</div>`;
  }
  if ((info.confusables || []).length) {
    html += `<h4>not to be confused with</h4><ul>` +
      info.confusables.map((c) => `<li><b>${esc(c.word)}</b> — ${esc(c.difference)}</li>`).join("") +
      `</ul>`;
  }
  if (info.memory_hook) html += `<h4>to remember it</h4><p>${esc(info.memory_hook)}</p>`;

  if (!html) return "";
  return `<div class="why">${html}<h4>from ${esc(from)}</h4></div>`;
}
